src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala

```diff
@@ -268,40 +329,86 @@ class InstanceOpFactoryImpl(
+        // labeled with a taskId that does not relate to a task existing in Mesos (previously, Marathon reused taskIds so
+        // there was always a 1:1 correlation from reservation to taskId)
+        val taskId = Task.Id.forRunSpec(runSpec.id)
+        val reservationLabels = TaskLabels.labelsForTask(frameworkId, taskId)
+        val reservation = Task.Reservation(persistentVolumeIds, Task.Reservation.State.New(timeout = Some(timeout)))
+        val task = Task.Reserved(
+          taskId = taskId,
+          reservation = reservation,
+          status = Task.Status(
+            stagedAt = now,
+            condition = Condition.Reserved,
+            networkInfo = networkInfo
+          ),
+          runSpecVersion = runSpec.version
+        )
+        val instance = Instance(
+          instanceId = task.taskId.instanceId,
+          agentInfo = agentInfo,
+          state = InstanceState(
+            condition = Condition.Reserved,
+            since = now,
+            activeSince = None,
+            healthy = None
+          ),
+          tasksMap = Map(task.taskId -> task),
+          runSpecVersion = runSpec.version,
+          unreachableStrategy = runSpec.unreachableStrategy
+        )
+        val stateOp = InstanceUpdateOperation.Reserve(instance)
+        (reservationLabels, stateOp)
+
+      case pod: PodDefinition =>
+        val instanceId = Instance.Id.forRunSpec(runSpec.id)
+        val taskIds = pod.containers.map { container =>
+          Task.Id.forInstanceId(instanceId, Some(container))
+        }
+
+        val networkInfos = podTaskNetworkInfos(pod, agentInfo, taskIds, resourceMatch.hostPorts)
+        val reservationLabels = TaskLabels.labelsForTask(
+          frameworkId,
+          taskIds.headOption.getOrElse(throw new IllegalStateException("pod does not have any container")))
+        val reservation = Task.Reservation(persistentVolumeIds, Task.Reservation.State.New(timeout = Some(timeout)))
+
+        val tasks = taskIds.map { taskId =>
+          Task.Reserved(
+            taskId = taskId,
+            reservation = reservation,
+            status = Task.Status(
+              stagedAt = now,
+              condition = Condition.Reserved,
+              networkInfo = networkInfos.getOrElse(
+                taskId, throw new Exception("failed to retrieve a task network info"))
+            ),
+            runSpecVersion = runSpec.version
+          )
+        }
+        val instance = Instance(
+          instanceId = instanceId,
+          agentInfo = agentInfo,
+          state = InstanceState(
+            condition = Condition.Reserved,
+            since = now,
+            activeSince = None,
+            healthy = None
+          ),
+          tasksMap = tasks.map(t => t.taskId -> t)(collection.breakOut),
+          runSpecVersion = runSpec.version,
+          unreachableStrategy = runSpec.unreachableStrategy
+        )
+        val stateOp = InstanceUpdateOperation.Reserve(instance)
```

Comment:
the part above looks pretty similar to the app section, can we generalize things here?

Reply:
I do not see any obvious advantage of generalizing things here. In my opinion, it will save  a couple of lines, but will hurt readability.